Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch

```diff
@@ -0,0 +1,594 @@
++    @NotNull
++    public Material getBannerWall() {
++        if (this == ColorType.NONE) {
++            return DyeableMaterial.BANNER_WALL.getDefaultMaterial();
++        } else {
++            return Objects.requireNonNull(Material.matchMaterial(this.name() + "_BANNER_WALL"));
++        }
++    }
++
++    /**
++     * Gets this {@link ColorType} color as an colored Dye-{@link Material}
++     * @return the colored Dye of this {@link ColorType}
++     */
++    @NotNull
++    public Material getDye() {
++        if (this == ColorType.NONE) {
++            return DyeableMaterial.DYE.getDefaultMaterial();
++        } else {
++            return Objects.requireNonNull(Material.matchMaterial(this.name() + "_DYE"));
++        }
++    }
++
++    @Override
++    public String toString() {
++        return "" + getChatColor();
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/item/DyeableMaterial.java b/src/main/java/io/papermc/paper/item/DyeableMaterial.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..cdfc0f3c2d16fca92abfc74d89850c16b5170ea4
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/item/DyeableMaterial.java
+@@ -0,0 +1,100 @@
++package io.papermc.paper.item;
++
++import org.bukkit.Material;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.Arrays;
++import java.util.Set;
++import java.util.function.Function;
++import java.util.stream.Collectors;
++
++/**
++ * Represents groups of different {@link Material}s, that can have different kind of colors ({@link ColorType}s)
++ */
++public enum DyeableMaterial {
++
++    WOOL(ColorType::getWool),
++    TERRACOTTA(ColorType::getTerracotta),
++    CONCRETE(ColorType::getConcrete),
++    CONCRETE_POWDER(ColorType::getConcretePowder),
++    CARPET(ColorType::getCarpet),
++    STAINED_GLASS(ColorType::getStainedGlass),
++    STAINED_GLASS_PANE(ColorType::getStainedGlassPane),
++    SHULKER_BOX(ColorType::getShulkerBox),
++    GLAZED_TERRACOTTA(ColorType::getGlazedTerracotta),
++    BED(ColorType::getBed),
++    BANNER(ColorType::getBanner),
++    BANNER_WALL(ColorType::getBannerWall),
++    DYE(ColorType::getDye);
++
++    public Function<ColorType, Material> colorToMaterialProcess;
++
++    DyeableMaterial(Function<ColorType, Material> colorToMaterialProcess) {
++        this.colorToMaterialProcess = colorToMaterialProcess;
++    }
++
++    /**
++     * Generates the colored {@link Material} using the current {@link DyeableMaterial} and
++     * the target-{@link ColorType}
++     * @param colorType the color, which the {@link Material} will become
++     * @return the colored {@link DyeableMaterial}
++     */
++    @NotNull
++    public Material withColor(@NotNull ColorType colorType) {
++        return colorToMaterialProcess.apply(colorType);
++    }
++
++    /**
++     * A {@link Set<Material>} of all possible colored {@link Material}-variants of this {@link DyeableMaterial} (material-group)
```

Comment:
JavaDoc links don't accept generic parameters, this is why the build is failing.

Reply:
Thanks for the help!